### PR TITLE
Add new processor selector parameters in cirq-google/engine run methods

### PR DESCRIPTION
**Pull request by jurruti**
This pull request (PR) adds two new parameters to the run methods of `EngineProgram`, `Engine`, and `EngineProcessor`: `run_name` and `device_configuration_name`. These parameters allow users to select a specific configuration for the processor when running the circuit. They are also added to the `init` function of `ProcessorSampler` (and the `get_sampler` methods within `Engine` and `EngineProcessor`).

Additionally, the run methods of `EngineProgram` and `Engine` now include a parameter called `processor_id`. This parameter can be used to specify a specific processor for the run, and will substitute the deprecated parameter `processor_ids` , because allowing multiple candidate processors was not useful in practice.

### CODE REVIEW

File: cirq-google/cirq_google/engine/engine_client.py [hunk 1]
```diff
@@ -395,14 +395,16 @@ async def create_job_async(
         run_name: str = "",
         device_config_name: str = "",
     ) -> Tuple[str, quantum.QuantumJob]:
-        """Creates and runs a job on Quantum Engine.
+        """Creates and runs a job on Quantum Engine. Either both `run_name` and
```
Comment (verult) [inline cirq-google/cirq_google/engine/engine_client.py:398]: Following the Python docstring conventions, the first line of docstring should be a one-line summary, followed by a blank line, and the more elaborate description: https://peps.python.org/pep-0257/#multi-line-docstrings

Ditto elsewhere

File: cirq-google/cirq_google/engine/processor_sampler_test.py [hunk 2]
```diff
@@ -24,15 +24,29 @@
 @pytest.mark.parametrize('circuit', [cirq.Circuit(), cirq.FrozenCircuit()])
 def test_run_circuit(circuit):
     processor = mock.create_autospec(AbstractProcessor)
-    sampler = cg.ProcessorSampler(processor=processor)
+    run_name = "RUN_NAME"
```
Comment (verult) [inline cirq-google/cirq_google/engine/processor_sampler_test.py:27]: Add tests for the case where run_name and device_config_name are empty. Ditto for other `ProcessorSampler` methods.

File: cirq-google/cirq_google/engine/engine.py [hunk 3]
```diff
@@ -219,6 +219,10 @@ def run(
         program_labels: Optional[Dict[str, str]] = None,
         job_description: Optional[str] = None,
         job_labels: Optional[Dict[str, str]] = None,
+        *,
+        processor_id: str = "",
```
Comment (verult) [inline cirq-google/cirq_google/engine/engine.py:223]: Would it make sense to add a TODO for each user-facing method to deprecate `processor_ids`?

File: cirq-google/cirq_google/engine/engine.py [hunk 4]
```diff
@@ -236,19 +240,34 @@ def run(
...
+                available qubits, couplers, and supported gates in the processor.
+                If specified, `processor_id` is required to be set.
 
         Returns:
             A single Result for this run.
 
         Raises:
             ValueError: If no gate set is provided.
+            ValueError: If neither `processor_id` or `processor_ids` are set.
+            ValueError: If  only one of `run_name` and `device_config_name` are specified.
```
Comment (verult) [inline cirq-google/cirq_google/engine/engine.py:267]: nit: Extra space after "If"

File: cirq-google/cirq_google/engine/engine_processor.py [hunk 5]
```diff
@@ -237,6 +263,8 @@ async def run_sweep_async(
         program_labels: Optional[Dict[str, str]] = None,
         job_description: Optional[str] = None,
         job_labels: Optional[Dict[str, str]] = None,
+        run_name: str = "",
```
Comment (verult) [inline cirq-google/cirq_google/engine/engine_processor.py:266]: Do we want to update calibration methods?

**Review by verult — CHANGES_REQUESTED** [review]
The user-facing docstrings look great!

Should there be test changes for Engine, EngineProcessor, and EngineProgram?

File: cirq-google/cirq_google/engine/engine_processor.py (hunk 5 above)
Comment (jurruti) [reply cirq-google/cirq_google/engine/engine_processor.py:266]: We don't. Note that these arguments are for run_sweep, not run_calibration (github view is a bit confusing).

**Review by wcourtney — COMMENTED** [review]
LGTM % a couple of comments. Will defer to @verult to approve.

File: cirq-google/cirq_google/engine/abstract_processor.py [hunk 6]
```diff
@@ -64,6 +64,8 @@ async def run_async(
         program_labels: Optional[Dict[str, str]] = None,
         job_description: Optional[str] = None,
         job_labels: Optional[Dict[str, str]] = None,
+        run_name: str = "",
```
[the remark this answers is not in the document]
Comment (jurruti) [reply cirq-google/cirq_google/engine/abstract_processor.py:67]: Whoops, thank you for the catch.

[the remark this answers is not in the document]
Comment (jurruti) [reply cirq-google/cirq_google/engine/abstract_processor.py:67]: I followed the decision we made in EngineClient to keep it as a String field instead of Optional[String] because there is no semantic difference between "" and None.

However, I do think it would be good to make them all Optional to have more consistency. I can do it in a different PR if that is ok.

**Review by verult — APPROVED** [review]
Congrats on wrapping up the project!!!
